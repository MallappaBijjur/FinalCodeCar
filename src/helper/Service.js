import axios from 'axios';

export const getCars = path => {
	return new Promise((res, rej) => {
		return axios.get(`https://auto1-mock-server.herokuapp.com/api/cars${path ? path: ''}`)
			.then(data => res(data))
			.catch(err => {
				rej(err);
			});
	});
}

export const getColors = path => {
	return new Promise((res, rej) => {
		return axios.get("https://auto1-mock-server.herokuapp.com/api/colors")
			.then(data => res(data))
			.catch(err => {
				rej(err);
			});
	});
}

export const getManufacturers = path => {
	return new Promise((res, rej) => {
		return axios.get("https://auto1-mock-server.herokuapp.com/api/manufacturers")
			.then(data => res(data))
			.catch(err => {
				rej(err);
			});
	});
}


export const getOneCarInfo = stockNumber => {
	return new Promise((res, rej) => {
		return axios.get(`https://auto1-mock-server.herokuapp.com/api/cars/${stockNumber}`)
			.then(data => res(data))
			.catch(err => {
				rej(err);
			});
	});
}

