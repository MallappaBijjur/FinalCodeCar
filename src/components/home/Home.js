import React, { useState, useEffect } from 'react';
import Filter from '../filter/Filter';
import CarsLanding from '../car/CarsLanding';
import TopHeader from '../common/TopHeader';
import Footer from '../common/Footer';
import '../home-car.css';
import { getCars, getColors, getManufacturers } from '../../helper/Service';
import { All_COLORS, ALL_MANUFACTURERS } from '../../constants';
import { getApiPath } from '../../utils';

const Home = () =>  {
	const [state, setState] = useState({ cars: [],colors: [], manufacturers: [], totalPageCount: 0, totalCarsCount: 0, currentPageNo: 1, selectedColor: All_COLORS, selectedManufacturer: ALL_MANUFACTURERS });
     
	useEffect(() => {
		const path = getApiPath({ selectedManufacturer: ALL_MANUFACTURERS, selectedColor: All_COLORS, page: 1})
		Promise.all([getCars(path), getColors(), getManufacturers()]).then(values => {
			const { data: carsData} = values[0];
			const { data: colorsData} = values[1];
			const { data: manufacturersData} = values[2];
			setState({ ...carsData, ...colorsData, ...manufacturersData, currentPageNo: 1, selectedColor: All_COLORS, selectedManufacturer: ALL_MANUFACTURERS });
		});
		
	}, []);

	const handleFilter = filterInfo => {
		const path = getApiPath({selectedManufacturer: filterInfo.selectedManufacturer, selectedColor: filterInfo.selectedColor});
		 getCars(path).then(({data}) => {
		 	setState({...state, ...data, selectedColor: filterInfo.selectedColor , selectedManufacturer: filterInfo.selectedManufacturer});
		 });
	}

	const getCarsPerPage = pageInfo => {
		const path = getApiPath({ selectedManufacturer: state.selectedManufacturer, selectedColor: state.selectedColor, page: pageInfo });
		 getCars(path).then(({data}) => {
		 	setState({ ...state, ...data, currentPageNo: pageInfo });
		 });
	}


	return(
		<main className="wrap-container" data-testid="home-main">
			<TopHeader />
			<section className='home-container'>
				<Filter
					colors={state.colors}
					manufacturers={state.manufacturers}
					handleFilter={handleFilter}
				/>
				<CarsLanding 
					cars={state.cars}
					totalCarsCount={state.totalCarsCount}
					totalPageCount={state.totalPageCount}
					currentPageNo={state.currentPageNo}
					getCarsPerPage={getCarsPerPage}
				/>
			</section>
			<Footer />
		</main>
	)
}

export default Home;