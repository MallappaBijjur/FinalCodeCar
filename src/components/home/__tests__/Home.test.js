import React from 'react';
import ReactDOM from 'react-dom';
import { render, fireEvent, cleanup, waitForElement } from '@testing-library/react';
import Home from '../Home';
import rerender from 'react-test-renderer';
import { getCars, getColors, getManufacturers } from '../../../helper/Service';

let mockPromiseCar;
let mockPromiseColors;
let mockPromiseManufacturers;

jest.mock('../../../helper/Service', () =>({
	getCars: jest.fn().mockImplementation(() => mockPromiseCar),
	getColors: jest.fn().mockImplementation(() => mockPromiseColors), 
	getManufacturers: jest.fn().mockImplementation(() => mockPromiseManufacturers)
}));

describe("Home", () => {

	const originalError = console.error
	beforeAll(() => {
		console.error = (...args) => {
			if (/Warning.*not wrapped in act/.test(args[0])) {
				return
			}
			originalError.call(console, ...args)
		}
	})

	afterAll(() => {
		console.error = originalError
	});

	afterEach(cleanup);

	beforeEach(() => {
		getCars.mockClear();
		getColors.mockClear();
		getManufacturers.mockClear();
	});

	
	const carsData = {
		"cars":[
			 {
					"stockNumber":30562,
					"manufacturerName":"Fiat",
					"modelName":"Ducato",
					"color":"yellow",
					"mileage":{
						 "number":100013,
						 "unit":"km"
					},
					"fuelType":"Petrol",
					"pictureUrl":"https://auto1-js-task-api--mufasa71.repl.co/images/car.svg"
			 },
			 {
					"stockNumber":35592,
					"manufacturerName":"Chrysler",
					"modelName":"New Yorker",
					"color":"white",
					"mileage":{
						 "number":100115,
						 "unit":"km"
					},
					"fuelType":"Petrol",
					"pictureUrl":"https://auto1-js-task-api--mufasa71.repl.co/images/car.svg"
			 }]
	};
	const fakeColorData = {"colors":["red","blue","green","black","yellow","white","silver"]};
	const fakeManufacturersData = {
		"manufacturers":[
			 {
					"name":"Audi",
					"models":[]
			 },
			 {
					"name":"BMW",
					"models":[]
			 },
			 {
					"name":"Chrysler",
					"models":[]
			 }]
			};

	it("renders without crashing", ()=>{
		mockPromiseManufacturers= Promise.resolve({ data:[] });
		mockPromiseCar= Promise.resolve({ data:[] });
		mockPromiseColors= Promise.resolve({ data:[] });
		const div = document.createElement("div");
		ReactDOM.render(<Home />, div);
	});
	
	it('renders Home component', () => {
		mockPromiseManufacturers= Promise.resolve({ data:[] });
		mockPromiseCar= Promise.resolve({ data:[] });
		mockPromiseColors= Promise.resolve({ data:[] });
		const { queryByTestId } = render(<Home />);   
		const linkElement = queryByTestId(/home-main/i);
		expect(linkElement).toBeInTheDocument();
	});
	
	it("matches snapshot", ()=> {
		mockPromiseManufacturers= Promise.resolve({ data:[] });
		mockPromiseCar= Promise.resolve({ data:[] });
		mockPromiseColors= Promise.resolve({ data:[] });
		const tree = rerender.create(<Home />).toJSON();
		expect(tree).toMatchSnapshot();
	});
	
	it("should render with api data", ()=> {

		getCars.mockImplementationOnce(() => {
			return Promise.resolve({
				data: carsData
			})
		});
		
		getManufacturers.mockImplementationOnce(() => {
			return Promise.resolve({
				data: fakeManufacturersData
			})
		})
		
		getColors.mockImplementationOnce(() => {
			return Promise.resolve({
				data: fakeColorData
			})
		});

		render(<Home />);
		expect(getCars).toHaveBeenCalledTimes(1);
		expect(getColors).toHaveBeenCalledTimes(1);
		expect(getManufacturers).toHaveBeenCalledTimes(1);
	});

	//TODO 
	it("should test filter functionality", async ()=> {
		
		getCars.mockImplementationOnce(() => {
			return Promise.resolve({
				data: carsData
			})
		});
		
		getManufacturers.mockImplementationOnce(() => {
			return Promise.resolve({
				data: fakeManufacturersData
			})
		})
		
		getColors.mockImplementationOnce(() => {
			return Promise.resolve({
				data: fakeColorData
			})
		});

		const { getByTestId } = render(<Home />);
			fireEvent.click(getByTestId("filter"), { target: { value: { manufacturerName: "Fiat", selectedColor: ""}}});
			const car = await waitForElement (() =>getByTestId("car-30562"));
			const carItems = getByTestId("cars-items");
			expect(car).toHaveTextContent(/View details/i);
			expect(carItems.children.length).toBe(2);
	});
});




