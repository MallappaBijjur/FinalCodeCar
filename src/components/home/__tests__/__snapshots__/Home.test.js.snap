// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home matches snapshot 1`] = `
<main
  className="wrap-container"
  data-testid="home-main"
>
  <header
    className="header"
  >
    <div>
      <img
        alt="Auto1LOgo"
        className="header__logo"
        src="auto1logo.jpg"
      />
    </div>
    <div
      className="header__nav"
      data-testid="header"
    >
      <a
        className="header__navlink"
        href="#purchase"
      >
        Purchase
      </a>
      <a
        className="header__navlink"
        href="#myOrders"
      >
        My Orders
      </a>
      <a
        className="header__navlink"
        href="#Sell"
      >
        Sell
      </a>
    </div>
  </header>
  <section
    className="home-container"
  >
    <section
      className="filter"
      data-testid="filter-comp"
    >
      <section
        className="filter-box"
      >
        <p>
          Color
        </p>
        <div
          className="dropdown"
          onKeyDown={[Function]}
        >
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="dropdown-white dropdown-toggle btn btn-primary"
            disabled={false}
            id="dropdown-basic"
            onClick={[Function]}
            type="button"
          >
            All car colors
          </button>
        </div>
        <p>
          Manufacturer
        </p>
        <div
          className="dropdown"
          onKeyDown={[Function]}
        >
          <button
            aria-expanded={false}
            aria-haspopup={true}
            className="dropdown-white dropdown-toggle btn btn-primary"
            disabled={false}
            id="dropdown-basic"
            onClick={[Function]}
            type="button"
          >
            All manufacturers
          </button>
        </div>
        <button
          className="filter-button btn btn-primary"
          data-testid="filter"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Filter
        </button>
      </section>
    </section>
    <div
      className="landing"
    >
      <div
        className="landing__top"
      >
        <p
          className="landing__top--text--bold"
          id="available cars"
        >
          Available Cars
        </p>
        <p
          className="landing__top--info"
        >
          Showing 0 of 0 results
        </p>
      </div>
      <div
        data-testid="cars-items"
      />
      <div
        className="pagination"
      >
        <a
          className="link__item--text link__item--text--small pagelink__text--orange"
          href="#available cars"
          onClick={[Function]}
        >
          First
        </a>
        <a
          className="link__item--text link__item--text--small pagelink__text--orange"
          href="#available cars"
          onClick={[Function]}
        >
          Previous
        </a>
        <a
          className="link__item--text link__item--text--small link__text--black"
          href="#available cars"
        >
          Page 1 of 0
        </a>
        <a
          className="link__item--text link__item--text--small pagelink__text--orange"
          href="#available cars"
          onClick={[Function]}
        >
          Next
        </a>
        <a
          className="link__item--text link__item--text--small pagelink__text--orange"
          href="#available cars"
          onClick={[Function]}
        >
          Last
        </a>
      </div>
    </div>
  </section>
  <footer
    className="footer"
  >
    <p
      className="footer__text"
    >
      Â© AUTO1 Group 2018
    </p>
  </footer>
</main>
`;
